mmdet/models/backbones/darknet.py

```diff
@@ -0,0 +1,181 @@
+        act_cfg (dict): Config dict for activation layer.
+            Default: dict(type='LeakyReLU', negative_slope=0.1).
+        norm_eval (bool): Whether to set norm layers to eval mode, namely,
+            freeze running stats (mean and var). Note: Effect on Batch Norm
+            and its variants only.
+
+    Example:
+        >>> from mmdet.models import Darknet
+        >>> import torch
+        >>> self = Darknet(depth=53)
+        >>> self.eval()
+        >>> inputs = torch.rand(1, 3, 416, 416)
+        >>> level_outputs = self.forward(inputs)
+        >>> for level_out in level_outputs:
+        ...     print(tuple(level_out.shape))
+        ...
+        (1, 256, 52, 52)
+        (1, 512, 26, 26)
+        (1, 1024, 13, 13)
+    """
+
+    def __init__(self,
+                 depth=53,
+                 out_indices=(3, 4, 5),
+                 norm_cfg=dict(type='BN', requires_grad=True),
+                 act_cfg=dict(type='LeakyReLU', negative_slope=0.1),
+                 norm_eval=True):
+        super(Darknet, self).__init__()
+        self.depth = depth
+        self.out_indices = out_indices
+        if self.depth == 53:
+            self.layers = [1, 2, 8, 8, 4]
+            self.channels = [[32, 64], [64, 128], [128, 256], [256, 512],
+                             [512, 1024]]
+        else:
+            raise KeyError(f'invalid depth {depth} for darknet')
+
+        cfg = dict(norm_cfg=norm_cfg, act_cfg=act_cfg)
+
+        self.conv1 = ConvModule(3, 32, 3, padding=1, **cfg)
+
+        self.cr_blocks = ['conv1']
+        for i, n_layers in enumerate(self.layers):
+            layer_name = f'cr_block{i + 1}'
+            in_c, out_c = self.channels[i]
+            self.add_module(
+                layer_name,
+                make_conv_and_res_block(in_c, out_c, n_layers, **cfg))
+            self.cr_blocks.append(layer_name)
+
+        self.norm_eval = norm_eval
+
+    def forward(self, x):
+        outs = []
+        for i, layer_name in enumerate(self.cr_blocks):
+            cr_block = getattr(self, layer_name)
+            x = cr_block(x)
+            if i in self.out_indices:
+                outs.append(x)
+
+        return tuple(outs)
+
+    def init_weights(self, pretrained=None):
+        if isinstance(pretrained, str):
+            logger = logging.getLogger()
+            load_checkpoint(self, pretrained, strict=False, logger=logger)
+        elif pretrained is None:
+            for m in self.modules():
+                if isinstance(m, nn.Conv2d):
+                    kaiming_init(m)
+                elif isinstance(m, (_BatchNorm, nn.GroupNorm)):
+                    constant_init(m, 1)
+
+        else:
+            raise TypeError('pretrained must be a str or None')
+
+    def _freeze_stages(self):
+        for param in self.parameters():
+            param.requires_grad = False
+
+    def train(self, mode=True):
+        super(Darknet, self).train(mode)
+        self._freeze_stages()
```

Comment:
We may make freeze optional. 
Btw, are all parameters in YOLOv3 backbone frozen in the original setting? If so, we may add comments about this.

Reply:
I'm sorry that I don't know if the original setting freezes all params in backbone, but it's unlikely I think. 

Also, according to https://github.com/ultralytics/yolov3/issues/998, freezing layers always result in worse performance. So, do you think it makes sense that I add `frozen_stages` parameter to darknet backbone, just like the one in resnet?

Reply:
Yes. 

So are you training with backbone frozen or not?

Reply:
Yes, I was training with backbone frozen. So, shall we try training with backbone unfrozen? AFAIK, `gluon-cv`'s backbone is not frozen during training.

Reply:
It's weird that the performance with backbone frozen looks ok. We may first add `fronzen_stages` arguments and see what will happen if `frozen_stages == -1`.

Reply:
Okay, I get your point. I think it's because I loaded the backbone's pretrained weights, provided by @wuhy08 . I don't know if the weights he provided is converted from `https://pjreddie.com/darknet/imagenet/`, or is just extracted from the YOLOv3 model trained on COCO.

If it's just extracted from the full YOLOv3 model trained on COCO, then, we can explain why the performance with backbone frozen looks ok.

I have tried to train with 1x schedule and backbone unfrozen, and got 0.318 mAP. Compared with the frozen one's 32.1, it's actually a bit lower.